### ISSUE TITLE
Free/Used/Reserved ports overview in column table

### ISSUE THREAD

**Issue by PieterL75** [netbox-community/netbox#11633]
### NetBox version

v3.3.10

### Feature type

Change to existing functionality

### Proposed functionality

I would like to see a summary of all available/occupied ports on the device list page.
This could be added as a new column that can be added using the 'configure table'

![image](https://user-images.githubusercontent.com/74899468/216028576-7516e2f0-64a9-4b94-9a1b-c13d2da680e5.png)

Additionally, I would like to have an extra field in the 'Filter' to look for devices with at least X unconnected ports.

### Use case

When provisioning a new environment, I need to pick the correct rack where to deploy it. This is mainly done by looking for the required network ports for that environment.
I have a custom report, but having this integrated (in a SavedFilter !!!) would be nice

**Comment by PieterL75** [comment]
I agree on that.. One other option is to create two column per port-type (interface, front, rear,...) with the 'occupied' and 'available' count?

Last option : the ability to create 'calculated' custom fields.. We could leverage a jinja2 template to process the object data and generate a output.. then I could write that field myself, and it is highly customizable

**Comment by jeremystretch** [comment]
This has some overlap with #6347, which will cache the number of each of type component associated with a device/VM. It doesn't consider whether each is occupied, but it would provide one of the two variables we need. With that in place, it might be feasible to add an individual column for each component type, which would annotate a count of occupied components and compare it to the total number.

I'm going to mark this as blocked by #6347 for now, though happy to discuss further.

**Comment by PieterL75** [comment]
What about the 'calculated custom field' ? 
A jinja2 template could be used to create a return value that is shown in the column. Then we can use that for any purpose.
ex: when displaying an IP address on an interface, I would like to also display the vlan id. 

That would give a great flexibility to what people want to see and how they want it.
Maybe some extra guidance on how to display things (progress bar, badge, ...) and you have a total customizable table view without any plugins or serverside coding

**Comment by jeremystretch** [comment]
A custom field wouldn't be any different; the data still needs to come from the database. It can't be cached, because the count of free/total ports can change at any time.

**Comment by PieterL75** [comment]
I agree on that, it might be slow, but it does will be able to create a table field at the convenience of the user.
And with #6347, there will be more direct-accessible counters to use.

This customfield is out of scope of my initial request, but rather an result of thinking how this can be customized.
I can create a separate FR for that

**Comment by coloHsq** [comment]
Hi everyone, I've actually done something similar to have a "per rack" summary of patch panels and PDUs available ports.

![Screenshot 2023-04-07 alle 09 42 52](https://user-images.githubusercontent.com/46020242/230566231-3e10ef51-5d3e-4d6e-ba74-febac0e87d1c.png)

All the count aggregations are done on demand, and for like 300 racks it runs fast enough to be usable.

The querysets I'm using for ports aggregations can probably be adapted for this request.
Let me know if you're interested.

**Comment by PieterL75** [comment]
@coloHsq How do you generate those fields ?
Is that through a plugin ?
Interesting to see that you got this working though :-)

**Comment by jeremystretch** [comment]
This is no longer blocked, however I still don't see a clear path to an implementation.

**Comment by coloHsq** [comment]
I'm starting to see a practical implementation for this one.
By now, the main problem is to make it flexible enough to be used for most of the use cases, but i've managed to solve a big part of this problem so that i can dynamically filter the elements taken for count aggregations, and this solution can be used with stock tables and filters.  
This would allow something like "show me all devices that have at least 5 enabled qsfp+ ports" (or whatever else filter conditions supported on models that are being counted), so one can define what and how to run the aggregations and save'em in a filter.
I still see this feature as best implemented in a specific plugin, as this relays on subqueries annotated on the main one, and it seems quite unlikely i'll be able to modify the core views querysets, unless we decide that this feature is worth to be placed in the core part.

**Comment by coloHsq** [comment]
Hi, it's been a while since I said I would release a plugin for this. 
but finally here it is : https://github.com/coloHsq/aggregatron.git

**Comment by jeremystretch** [comment]
I'm going to close this out as no detailed implementation has been agreed upon. Happy to reopen if someone would like to invest the time and effort in drafting a detailed proposal.
